### PR TITLE
fix(env): avoid unnecessary `process.env` modifications for consistent Node and Bun results

### PR DESCRIPTION
**Pull request by byCedric**
# BEFORE MERGING

- [x] Delete the custom workflow

# Why

Fixes ENG-10434

After diving deep into the environment variable handling, there are some unexplained anomalies when running the exact same setup in Bun and Node. This seems to be caused by not just a single thing, but by multiple things.

Bun | Node
--- | ---
Bun result from testplan | Node result from testplan

In the `@expo/env` code, we use `dotenv` to overwrite envvars, and custom overwrite logic to finalize the setup. Unfortunately, this uses two different methods to set the `process.env`:
- `process.env.SOMEVAR = value` ([within `dotenv`](https://github.com/motdotla/dotenv/blob/5d00dd9d1561059247eb31a2841cbc2e2b14f273/lib/main.js#L280))
- `process.env = { ... }` ([within `@expo/env`](https://github.com/expo/expo/blob/6319bba18ccc9743b6432df181f1197ead27908e/packages/%40expo/env/src/env.ts#L139))

The 2nd method seems to [break down completely in Bun](https://twitter.com/cedric_dev/status/1724544717147296070), while Node does end up with the expected result.

## TLDR

When using a more complex `.env` setup (e.g. `.env`, `.env.development`, and `.env.local`), Bun and Node produce a different `process.env` setup.

# How

(bold are the most important ones)

- Used a temporary workflow to assert anomaly _(we can't run our tests with Bun easily)_
- Fixed environment variables leaking across tests in `@expo/env`
- Added environment variables expansion tests
- Added `NODE_ENV=test` tests
- **Parsed result of all dotenv files without writing to `process.env`**
- **Only expand variables once before applying envvars, to avoid unexpected overwrites in this step**
- **Only use `process.env[key] = value` to apply envvars**

# Test Plan

- `$ bun create expo ./test-env`
- `$ cd ./test-env`
- Create a **metro.config.js** with the content from below
- Create **.env**, **.env.development**, and **.env.local** from below

The result for these 4 different commands _**MUST**_ be the _**exact**_ same.
- `$ bun expo start`
- `$ yarn expo start`
- `$ npx expo start`
- `$ pnpm expo start`

### Test files

<details><summary><strong>metro.config.json</strong> content</summary>

```js
const { getDefaultConfig } = require('expo/metro-config');

console.log('ENVVARS');
console.log('TEST_OVERWRITE', process.env.TEST_OVERWRITE);
console.log('TEST_.env.local', process.env['TEST_.env.local']);
console.log('TEST_.env.development', process.env['TEST_.env.development']);
console.log('TEST_.env', process.env['TEST_.env']);

process.exit(0);

module.exports = getDefaultConfig(__dirname);
```

</details>

<details><summary><strong>.env</strong> content</summary>

```
TEST_OVERWRITE=.env
TEST_.env=loaded
```

</details>

<details><summary><strong>.env.development</strong> content</summary>

```
TEST_OVERWRITE=.env.development
TEST_.env.development=loaded
```

</details>

<details><summary><strong>.env.local</strong> content</summary>

```
TEST_OVERWRITE=.env.local
TEST_.env.local=loaded
```

</details>

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by byCedric** [comment]
As expected, the anti-leak test is failing: https://github.com/expo/expo/actions/runs/6877416721/job/18705001953?pr=25393#step:6:84

This means that our tests are not accurate. It's basically writing to the `originalEnv`, causing environment variables to leak across tests. Just updated the test to fix that.

File: packages/@expo/env/src/env.ts [hunk 1]
```diff
@@ -136,7 +160,11 @@ export function createControlledEnvironment() {
       }
     }
 
-    process.env = { ...process.env, ...envInfo.env };
+    for (const key of Object.keys(envInfo.env)) {
+      // Avoid creating a new object, mutate it instead as this causes problems in Bun
```
Comment (brentvatne) [inline packages/@expo/env/src/env.ts:164]: 😵‍💫

Comment (byCedric) [reply packages/@expo/env/src/env.ts:164]: I mean, do you want me to add a link to my tweet as comment? https://twitter.com/cedric_dev/status/1724544717147296070 🙈

**Review by tsapeta — APPROVED** [review]
